src/services/game-logic/GameEngine.ts

```diff
@@ -12,13 +13,58 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let legalMoves = piece?.getPossibleMoves(this.board).filter(this.isOnBoard) ?? [];
+        let legalMoves = piece?.getPossibleMoves(this.board) ?? [];
         if (piece?.name === PieceNames.KING) {
             legalMoves = legalMoves.filter(
                 (move) => !this.isCastling(move, piece) || this.isCastlingLegal(move, piece)
             );
         }
-        return legalMoves.filter((destination) => !this.isOccupiedBySameColor(destination, piece));
+        return legalMoves
+            .filter((destination) => !this.isOccupiedBySameColor(destination, piece))
+            .filter((move) => this.moveNotResultWithCheck(move, piece, square));
+    };
+
+    moveNotResultWithCheck(move: Square, piece: Piece | null, square: Square) {
+        const potentialMove: Square = { row: move.row, column: move.column };
+        const potentialPiece = this.board.getPiece(potentialMove);
+
+        this.movePiece(square, potentialMove);
+        const isMovePossible = !this.isCheck(piece);
+
+        this.movePiece(potentialMove, square);
+        piece!.hasMoved = false;
+
+        if (potentialPiece)
+            this.board.state[potentialPiece.position.row][potentialPiece.position.column] = potentialPiece;
+
+        return isMovePossible;
+    }
+
+    private isCheck(piece: Piece | null): Boolean {
+        const color = piece?.color === Colors.BLACK ? Colors.WHITE : Colors.BLACK;
+        let currentPlayerKing: Square | undefined = this.getKingForCheck(piece)?.position;
+        const isChecked = this.isKingUnderCheck(color, currentPlayerKing);
+        return isChecked;
+    }
+
+    isKingUnderCheck(color: Colors, piecePosition: Square | undefined): boolean {
+        let checked = false;
+        this.board.checkAllSquares((square: Piece) => {
+            if (square && square.color === color) {
+                checked = square
+                    .getPossibleMoves(this.board)
+                    .some((move) => move.row === piecePosition?.row && move.column === piecePosition.column);
+            }
+        });
+        return checked;
+    }
+
+    getKingForCheck = (piece: Piece | null): King | null => {
+        let king = null;
+        this.board.checkAllSquares((square: Piece) => {
+            if (square && square?.color === piece?.color && square.name === PieceNames.KING) king = square;
```

Comment:
rozdzielmy to na oddzielne linie i opakujmy w { }

Reply:
poprawione